src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -0,0 +1,183 @@
+
+
+class IntrinsicCall(Call):
+    ''' Node representing a call to an intrinsic routine (function or
+    subroutine). This can be found as a standalone statement
+    or an expression.
+
+    '''
+    # Textual description of the node.
+    _children_valid_format = "[DataNode]*"
+    _text_name = "IntrinsicCall"
+    _colour = "cyan"
+    _min_arg_count = 0
+    _max_arg_count = 0
+
+    #: The intrinsics that can be represented by this node.
+    Intrinsic = Enum('Intrinsic', [
+        'ALLOCATE', 'DEALLOCATE', 'RANDOM'
+        ])
+    #: Named tuple for describing the properties of the required arguments to
+    #: a particular intrinsic. If there's no limit on the number of arguments
+    #: then `max_count` will be None.
+    ArgDesc = namedtuple('ArgDesc', 'min_count max_count types')
+    #: List of ArgDesc objects describing the required arguments, indexed
+    #: by intrinsic name.
+    _required_args = {}
+    #: Dict of optional arguments, indexed by intrinsic name. Each optional
+    #: argument is described by an entry in the Dict.
+    _optional_args = {}
+    # The positional arguments to allocate must all be References (or
+    # ArrayReferences but they are a subclass of Reference).
+    _required_args[Intrinsic.ALLOCATE] = ArgDesc(1, None, Reference)
+    _optional_args[Intrinsic.ALLOCATE] = {"mold": Reference,
+                                          "stat": Reference}
+    _required_args[Intrinsic.DEALLOCATE] = ArgDesc(1, None, Reference)
+    _optional_args[Intrinsic.DEALLOCATE] = {"stat": Reference}
+    _required_args[Intrinsic.RANDOM] = ArgDesc(1, 1, Reference)
+    _optional_args[Intrinsic.RANDOM] = {}
+
+    @classmethod
+    def create(cls, intrinsic, arguments):
+        '''Create an instance of this class given the type of intrinsic and a
+        list of nodes (or name-and-node tuples) for its arguments. Any
+        named arguments *must* come after any required arguments.
+
+        :param intrinsic: the Intrinsic being called.
+        :type intrinsic: py:class:`psyclone.psyir.IntrinsicCall.Intrinsic`
+        :param arguments: the arguments to this routine, and/or \
+            2-tuples containing an argument name and the \
+            argument. Arguments are added as child nodes.
+        :type arguments: List[ \
+            Union[:py:class:`psyclone.psyir.nodes.DataNode`, \
+                  Tuple[str, :py:class:`psyclone.psyir.nodes.DataNode`]]]
+
+        :returns: an instance of this class.
+        :rtype: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+
+        :raises TypeError: if any of the arguments are of the wrong type.
+        :raises ValueError: if any optional arguments have incorrect names \
+            or if a positional argument is listed *after* a named argument.
+        :raises ValueError: if the number of supplied arguments is not valid \
+            for the specified intrinsic.
+
+        '''
+        if not isinstance(intrinsic, cls.Intrinsic):
```

Comment:
pylint says "Second argument of isinstance is not a type". Would `intrinsic not in cls.Intrinsic` work for Enums?

Reply:
That only works if `intrinsic` is a member of an Enumeration type. If it's of the wrong type then we get a type error. I'm suspecting that this is a bug in pylint as I found https://stackoverflow.com/questions/38393897/how-to-check-if-a-variable-is-of-type-enum-in-python and examined the input to isinstance:

    (Pdb) type(cls.Intrinsic)
    <class 'enum.EnumMeta'>
    (Pdb) type(Enum)
    <class 'enum.EnumMeta'>

and if I replace `cls.Intrinsic` with `Enum` in the call to `isinstance` then the pylint error goes away.

Reply:
I've fixed it by first checking that it is an `Enum` and then checking that it is `in cls.Intrinsic`. I've also renamed the `intrinsic` argument back to `routine` to keep pylint happy.